pkg/hubble/observer/namespace_manager.go

```diff
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Hubble
+
+package observer
+
+import (
+	"context"
+	"time"
+
+	"golang.org/x/exp/slices"
+
+	observerpb "github.com/cilium/cilium/api/v1/observer"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+var _ NamespaceManager = &namespaceManager{}
+
+const (
+	checkNamespaceAgeFrequency = 5 * time.Minute
+	namespaceTTL               = time.Hour
+)
+
+type NamespaceManager interface {
+	GetNamespaces() []*observerpb.Namespace
+	AddNamespace(...*observerpb.Namespace)
+}
+
+type namespaceRecord struct {
+	namespace *observerpb.Namespace
+	added     time.Time
+}
+
+type namespaceManager struct {
+	mu         lock.RWMutex
+	namespaces map[string]namespaceRecord
+	nowFunc    func() time.Time
+}
+
+func NewNamespaceManager() *namespaceManager {
+	return &namespaceManager{
+		namespaces: make(map[string]namespaceRecord),
+		nowFunc:    time.Now,
+	}
+}
+
+func (m *namespaceManager) Run(ctx context.Context) {
+	ticker := time.NewTicker(checkNamespaceAgeFrequency)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			// periodically remove any namespaces which haven't been seen in flows
+			// for the last hour
+			m.cleanupNamespaces()
+		}
+	}
+}
+
+func (m *namespaceManager) cleanupNamespaces() {
+	m.mu.Lock()
+	for key, record := range m.namespaces {
+		if record.added.Add(namespaceTTL).Before(m.nowFunc()) {
+			delete(m.namespaces, key)
+		}
+	}
+	m.mu.Unlock()
+}
+
+func (m *namespaceManager) GetNamespaces() []*observerpb.Namespace {
+	m.mu.RLock()
+	namespaces := make([]*observerpb.Namespace, 0, len(m.namespaces))
+	for _, ns := range m.namespaces {
+		namespaces = append(namespaces, ns.namespace)
+	}
```

Comment:
nit: no need for a loop here
```suggestion
	namespaces := make([]*observerpb.Namespace, len(m.namespaces))
	copy(namespaces, m.namespaces)
```

Reply:
Alex made the same comment, this isn't a slice we're coping from, but a map, so we cannot use `copy()`.

Reply:
Oops, my bad, sorry.